File: app/src/ultimate/java/com/clockmods/widget/render/WidgetThemeRegistry.java
package com.clockmods.widget.render;
import android.content.Context;
import android.content.res.Configuration;
import com.clockmods.R;
import com.clockmods.widget.model.*;
import java.util.*;
import static com.clockmods.widget.model.WidgetThemeSpec.Font.*;
public final class WidgetThemeRegistry {
    private WidgetThemeRegistry() { }
    public static List<WidgetThemeSpec> all(Context c) {
        boolean night=(c.getResources().getConfiguration().uiMode & Configuration.UI_MODE_NIGHT_MASK)==Configuration.UI_MODE_NIGHT_YES;
        int primary=c.getColor(night ? android.R.color.system_neutral1_50 : android.R.color.system_neutral1_900);
        int secondary=c.getColor(night ? android.R.color.system_neutral2_200 : android.R.color.system_neutral2_700);
        int accent=c.getColor(night ? android.R.color.system_accent1_200 : android.R.color.system_accent1_700);
        return Collections.unmodifiableList(Arrays.asList(
            new WidgetThemeSpec("system.dynamic",R.string.widget_theme_dynamic,R.drawable.widget_bg_dynamic,primary,secondary,accent,SANS,255),
            new WidgetThemeSpec("glass.light",R.string.widget_theme_glass,R.drawable.widget_bg_glass,0xff152a38,0xff304657,0xff125e70,SANS,235),
            new WidgetThemeSpec("instrument.dark",R.string.widget_theme_instrument,R.drawable.widget_bg_instrument,0xfff4f6f9,0xffacbac7,0xffa8dcfa,MONOSPACE,255),
            new WidgetThemeSpec("paper.warm",R.string.widget_theme_paper,R.drawable.widget_bg_paper,0xff30281f,0xff675541,0xff9e362c,SERIF,255),
            new WidgetThemeSpec("neon.night",R.string.widget_theme_neon,R.drawable.widget_bg_neon,0xffecfaff,0xffacbde0,0xff5fffe0,MONOSPACE,255),
            new WidgetThemeSpec("transparent.clean",R.string.widget_theme_transparent,R.drawable.widget_bg_transparent,0xfffafafa,0xfffafafa,0xfffafafa,SANS,0)));
    }
    public static WidgetThemeSpec resolve(Context c,String id) {
        for(WidgetThemeSpec t:all(c)) if(t.id.equals(id)) return t;
        return all(c).get(0);
    }
}

File: app/src/ultimate/java/com/clockmods/widget/render/WidgetThemeRegistry.kt
package com.clockmods.widget.render

import android.content.Context
import android.content.res.Configuration
import com.clockmods.R
import com.clockmods.widget.model.WidgetThemeSpec
import com.clockmods.widget.model.WidgetThemeSpec.Font.MONOSPACE
import com.clockmods.widget.model.WidgetThemeSpec.Font.SANS
import com.clockmods.widget.model.WidgetThemeSpec.Font.SERIF

object WidgetThemeRegistry {
    @JvmStatic fun all(context: Context): List<WidgetThemeSpec> {
        val night = context.resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK ==
            Configuration.UI_MODE_NIGHT_YES
        val primary = context.getColor(if (night) android.R.color.system_neutral1_50 else android.R.color.system_neutral1_900)
        val secondary = context.getColor(if (night) android.R.color.system_neutral2_200 else android.R.color.system_neutral2_700)
        val accent = context.getColor(if (night) android.R.color.system_accent1_200 else android.R.color.system_accent1_700)
        return listOf(
            WidgetThemeSpec("system.dynamic", R.string.widget_theme_dynamic, R.drawable.widget_bg_dynamic, primary, secondary, accent, SANS, 255),
            WidgetThemeSpec("glass.light", R.string.widget_theme_glass, R.drawable.widget_bg_glass, 0xff152a38.toInt(), 0xff304657.toInt(), 0xff125e70.toInt(), SANS, 235),
            WidgetThemeSpec("instrument.dark", R.string.widget_theme_instrument, R.drawable.widget_bg_instrument, 0xfff4f6f9.toInt(), 0xffacbac7.toInt(), 0xffa8dcfa.toInt(), MONOSPACE, 255),
            WidgetThemeSpec("paper.warm", R.string.widget_theme_paper, R.drawable.widget_bg_paper, 0xff30281f.toInt(), 0xff675541.toInt(), 0xff9e362c.toInt(), SERIF, 255),
            WidgetThemeSpec("neon.night", R.string.widget_theme_neon, R.drawable.widget_bg_neon, 0xffecfaff.toInt(), 0xffacbde0.toInt(), 0xff5fffe0.toInt(), MONOSPACE, 255),
            WidgetThemeSpec("transparent.clean", R.string.widget_theme_transparent, R.drawable.widget_bg_transparent, 0xfffafafa.toInt(), 0xfffafafa.toInt(), 0xfffafafa.toInt(), SANS, 0),
        )
    }

    @JvmStatic fun resolve(context: Context, id: String?): WidgetThemeSpec =
        all(context).firstOrNull { it.id == id } ?: all(context).first()
}
